### PR TITLE
cmd/scollector: fastly collector

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

File: metadata/metadata.go [hunk 1]
```diff
@@ -118,6 +118,7 @@ const (
 	Thread               = "threads"
 	Timestamp            = "timestamp"
 	Transition           = "transitions"
+	USD                  = "us dollars"
```
Comment (alienth) [inline metadata/metadata.go]: nitpick: "US dollars", like "RPM"

File: cmd/scollector/collectors/fastly.go [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+			} else {
+				v = float64(value.(int64))
+			}
+			if shouldDiv {
+				v /= 60.0
+			}
+			AddTS(md, prefix+"."+metricName+suffix, timeStamp, v, ts, metadata.RateType(rateTag), metadata.Unit(unitTag), descTag)
+		case string:
+			// Floats in strings, I know not why, precision perhaps?
+			if f, err := strconv.ParseFloat(value, 64); err != nil {
```
Comment (alienth) [inline cmd/scollector/collectors/fastly.go]: Unhandled error. Seems like we should at least print this?

File: cmd/scollector/collectors/fastly.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+		}
+		metricName := jsonTag
+		if metricTag != "" {
+			metricName = metricTag
+		}
+		if metricName == "" {
+			slog.Errorf("Unable to determine metric name for field %s. Skipping.", field.Name)
+			continue
+		}
+		shouldDiv := divTag != ""
```
Comment (alienth) [inline cmd/scollector/collectors/fastly.go]: Weak check which may cause bugs if things get extended. Why not check to see if it is == 'true'?

File: cmd/scollector/collectors/fastly.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+	Status4xx          int64   `json:"status_4xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 4xx response code (Client Error)."`
+	Status503          int64   `json:"status_503" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 503 response code. (Service Unavailable)"`
+	Status5xx          int64   `json:"status_5xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a xxx response code."`
+	Synth              int64   `json:"synth"`
+	TLS                int64   `json:"tls"`
+	Uncacheable        int64   `json:"uncacheable" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests that were uncacheable."`
+	Video              int64   `json:"video"`
+}
+
+const fastlyDivDesc = " This metric is collected per minute, but we divide it by 60 seconds in order to normalize the rate to per second instead of per minute."
```
Comment (alienth) [inline cmd/scollector/collectors/fastly.go]: Intentional leading whitespace?

File: cmd/scollector/collectors/fastly.go [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+	Status302          int64   `json:"status_302" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 302 response code (Found)."`
+	Status304          int64   `json:"status_304" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 304 response code."`
+	Status3xx          int64   `json:"status_3xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 3xx response code (Redirection)."`
+	Status4xx          int64   `json:"status_4xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 4xx response code (Client Error)."`
+	Status503          int64   `json:"status_503" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a 503 response code. (Service Unavailable)"`
+	Status5xx          int64   `json:"status_5xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a xxx response code."`
+	Synth              int64   `json:"synth"`
```
Comment (alienth) [inline cmd/scollector/collectors/fastly.go]: Would add description: 'Synthetic response sent from VCL. Typically used to send edge-generated error pages.'

File: cmd/scollector/collectors/fastly.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+	}
+	return md, nil
+}
+
+type FastlyClient struct {
+	key    string
+	client *http.Client
+}
+
+func NewFastlyClient(key string) FastlyClient {
```
Comment (captncraig) [inline cmd/scollector/collectors/fastly.go]: don't export anything you don't have to.

File: cmd/scollector/conf/conf.go [hunk 7]
```diff
@@ -99,6 +100,10 @@ type GoogleAnalytics struct {
 	Sites    []GoogleAnalyticsSite
 }
 
+type Fastly struct {
```
Comment (captncraig) [inline cmd/scollector/conf/conf.go]: unexport

File: cmd/scollector/collectors/fastly.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+	Status5xx          int64   `json:"status_5xx" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests with a xxx response code."`
+	Synth              int64   `json:"synth"`
+	TLS                int64   `json:"tls"`
+	Uncacheable        int64   `json:"uncacheable" div:"true" rate:"gauge" unit:"requests per second" desc:"The number of http requests that were uncacheable."`
+	Video              int64   `json:"video"`
+}
+
+const fastlyDivDesc = " This metric is collected per minute, but we divide it by 60 seconds in order to normalize the rate to per second instead of per minute."
+
+func FastlyReflectAdd(md *opentsdb.MultiDataPoint, prefix, suffix string, st interface{}, timeStamp int64, ts opentsdb.TagSet) {
```
Comment (captncraig) [inline cmd/scollector/collectors/fastly.go]: unexport

File: cmd/scollector/conf/conf.go (hunk 7 above)
Comment (alienth) [inline cmd/scollector/conf/conf.go]: That's directly referenced in the collectors package.

Comment (captncraig) [inline cmd/scollector/conf/conf.go]: my bad on this one.

Comment (captncraig) [inline cmd/scollector/conf/conf.go]: should it live *in* the conf package then?

**Comment by captncraig** [comment]
lgtm if you unexport anything that doesn't need to be.

File: cmd/scollector/collectors/fastly.go (hunk 5 above)
Comment (kylebrandt) [inline cmd/scollector/collectors/fastly.go]: @alienth What would the unit be?

**Pull request merged**